File: src/main/java/org/devcon/ticket/CapabilityValidator.java
package org.devcon.ticket;

import com.alphawallet.attestation.core.SignatureUtility;
import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.interfaces.DecodedJWT;
import com.auth0.jwt.interfaces.Verification;
import java.net.MalformedURLException;
import java.net.URL;
import java.security.PublicKey;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;

/**
 * Verified a long-term JWT that has been issued to a specific domain in order to give it access to the DevCon ticket API.
 *
 * This is specifically going to be used to allow a third party site to open an iframe to ticket.devcon
 * in order to access the ticket secret stored in the local cache to construct a useDevconTicket request.
 */
public class CapabilityValidator extends CapabilityCommon{
  private final PublicKey verifyingKey;
  private final URL verifierDomain;
  private final Verification verification;

  public CapabilityValidator(AsymmetricKeyParameter verifyingKey, String verifierDomain) throws MalformedURLException {
    this.verifyingKey = SignatureUtility.convertPublicBouncyCastleKeyToJavaKey(verifyingKey);
    this.verifierDomain = new URL(verifierDomain);
    this.verification = JWT.require(getAlgorithm(this.verifyingKey, null))
        .withAudience(this.verifierDomain.toString())
        .withIssuer(this.verifierDomain.toString());
  }

  public boolean validateRequest(String jsonInput, String domain, Set<String> tasksThatMustBePresent) {
    try {
      URL domainUrl = new URL(domain);
      // Note that we already have added Audience, Issuer and that time validity and signature are
      // always verified implicitly.
      JWTVerifier verifier = verification.withSubject(domainUrl.toString()).build();
      DecodedJWT jwt = JWT.decode(jsonInput);
      jwt = verifier.verify(jwt);
      return verifyTasks(jwt, tasksThatMustBePresent);
    } catch (Exception e) {
      return false;
    }
  }

  private boolean verifyTasks(DecodedJWT jwt, Set<String> tasksThatMustBePresent) {
    String tasksString = jwt.getClaim(TasksClaimName).asString();
    Set<String> tasksInJwt = new HashSet<String>(Arrays.asList(tasksString.split(",")));
    Set<String> trimmedTasksThatMustBePresent = tasksThatMustBePresent.stream().
        map(s -> s.toLowerCase().trim()).collect(Collectors.toSet());
    return tasksInJwt.containsAll(trimmedTasksThatMustBePresent)
        && tasksThatMustBePresent.size() > 0;
  }

}

File: src/main/javascript/crypto/src/libs/CapabilityValidator.ts
import {CapabilityIssuer} from "./CapabilityIssuer";

const jwt = require('jsonwebtoken');
/**
 * Verified a long-term JWT that has been issued to a specific domain in order to give it access to the DevCon ticket API.
 *
 * This is specifically going to be used to allow a third party site to open an iframe to ticket.devcon
 * in order to access the ticket secret stored in the local cache to construct a useDevconTicket request.
 */
export class CapabilityValidator {
    private jwtData: any = {};

    constructor (private verifyingKey: string, private verifierDomain: string) {
        // let verification = jwt.require(getAlgorithm(this.verifyingKey, null))
        this.jwtData['audience'] = verifierDomain;
        this.jwtData['issuer'] = verifierDomain;
    }

    validateRequest(jsonInput: string, domain: string, tasksThatMustBePresent: string[]) {
        try {
            // Note that we already have added Audience, Issuer and that time validity and signature are
            // always verified implicitly.
            jwt.verify(jsonInput, this.verifyingKey, Object.assign(this.jwtData,{
                // https://github.com/auth0/node-jsonwebtoken
                //     algorithms: List of strings with the names of the allowed algorithms. For instance, ["HS256", "HS384"].
                // audience: if you want to check audience (aud), provide a value here. The audience can be checked against a string, a regular expression or a list of strings and/or regular expressions.
                //     Eg: "urn:foo", /urn:f[o]{2}/, [/urn:f[o]{2}/, "urn:bar"]
                //
                // complete: return an object with the decoded { payload, header, signature } instead of only the usual content of the payload.
                // issuer (optional): string or array of strings of valid values for the iss field.
                // jwtid (optional): if you want to check JWT ID (jti), provide a string value here.
                //     ignoreExpiration: if true do not validate the expiration of the token.
                //     ignoreNotBefore...
                // subject: if you want to check subject (sub), provide a value here
                // clockTolerance: number of seconds to tolerate when checking the nbf and exp claims, to deal with small clock differences among different servers
                // maxAge: the maximum allowed age for tokens to still be valid. It is expressed in seconds or a string describing a time span zeit/ms.
                //     Eg: 1000, "2 days", "10h", "7d". A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc), otherwise milliseconds unit is used by default ("120" is equal to "120ms").
                //
                // clockTimestamp: the time in seconds that should be used as the current time for all necessary comparisons.
                //     nonce: if you want to check nonce claim, provide a string value here. It is used on Open ID for the ID Tokens.
                subject: 'domain',
                issuer: 'urn:issuer'

            }), (err:any, decoded:any) => {
                // if issuer mismatch, err == invalid issuer
                return this.verifyTasks(decoded, tasksThatMustBePresent);
            });

            return false;

        } catch (e) {
            return false;
        }
    }

    verifyTasks(jwt: any, tasksThatMustBePresent: string[]) {
        let tasksString: string = jwt[CapabilityIssuer.TasksClaimName];
        let tasksInJwt: string[] = tasksString.split(",");
        return tasksThatMustBePresent.filter(task => !tasksInJwt.includes(task)).length == 0;
    }
}
